File: frontend/lib/schemas.ts
import { z } from "zod";

export const AthleteInfoSchema = z.object({
  athlete_id: z.string().nullable(),
  name: z.string().nullable(),
  location: z.string().nullable(),
  profile_url: z.string().nullable(),
});

export const AthleteListItemSchema = z.object({
  athlete_id: z.string(),
  name: z.string().nullable(),
  location: z.string().nullable(),
  activity_count: z.number(),
  profile_url: z.string().nullable(),
});

export const SportTypeOptionSchema = z.object({
  value: z.string(),
  label: z.string(),
  activity_type: z.string(),
});

export const MetaSchema = z.object({
  app_name: z.string(),
  app_subtitle: z.string(),
  unit_system: z.enum(["metric", "imperial"]),
  distance_unit: z.string(),
  elevation_unit: z.string(),
  sport_types: z.array(SportTypeOptionSchema),
  activity_count: z.number(),
  first_activity: z.string().nullable(),
  last_activity: z.string().nullable(),
  athlete: AthleteInfoSchema.nullable(),
  athletes: z.array(AthleteListItemSchema).default([]),
});

export const TotalsSchema = z.object({
  count: z.number(),
  distance: z.number(),
  elevation: z.number(),
  moving_time_s: z.number(),
  calories: z.number(),
});

export const ActivitySummarySchema = z.object({
  activity_id: z.string(),
  name: z.string(),
  start_date_time: z.string(),
  sport_type: z.string(),
  sport_label: z.string(),
  activity_type: z.string(),
  distance_km: z.number(),
  distance_mi: z.number(),
  elevation_m: z.number(),
  moving_time_s: z.number(),
  elapsed_time_s: z.number(),
  average_speed_kmh: z.number().nullable(),
  average_pace_s_per_km: z.number().nullable(),
  pace_unit: z.string(),
  average_heart_rate: z.number().nullable(),
  max_heart_rate: z.number().nullable(),
  average_power: z.number().nullable(),
  calories: z.number().nullable(),
  is_commute: z.boolean(),
  gear_name: z.string().nullable(),
  has_map: z.boolean(),
});

export const PaginatedActivitiesSchema = z.object({
  total: z.number(),
  limit: z.number(),
  offset: z.number(),
  items: z.array(ActivitySummarySchema),
});

const PeriodTotalsSchema = TotalsSchema.extend({ period: z.string() });
const LabeledTotalsSchema = TotalsSchema.extend({ label: z.string(), average_heart_rate: z.number().nullable().optional() });

const CalendarPointSchema = z.object({
  date: z.string(),
  count: z.number(),
  distance: z.number(),
  moving_time_s: z.number(),
  training_load: z.number(),
});

const EddingtonSummarySchema = z.object({
  activity_type: z.string(),
  number: z.number(),
  longest_day: z.number(),
  next: z.number(),
  days_to_next: z.number().nullable(),
});

const MilestoneSchema = z.object({
  achieved_on: z.string(),
  group: z.string(),
  title: z.string(),
  description: z.string(),
  icon: z.string(),
  sport_type: z.string().nullable(),
  activity_id: z.string().nullable(),
  fun_comparison: z.string().nullable(),
});

const GearItemSchema = z.object({
  gear_id: z.string(),
  name: z.string(),
  gear_type: z.string(),
  distance_km: z.number(),
  is_retired: z.boolean(),
});

export const DashboardSchema = z.object({
  empty: z.boolean(),
  filtered_empty: z.boolean().optional(),
  available_years: z.array(z.number()).optional(),
  unit_system: z.string(),
  totals: TotalsSchema.optional(),
  recent_activities: z.array(ActivitySummarySchema).optional(),
  weekly_stats: z.array(PeriodTotalsSchema).optional(),
  monthly_stats: z.array(PeriodTotalsSchema).optional(),
  yearly_stats: z.array(PeriodTotalsSchema).optional(),
  activity_calendar: z.array(CalendarPointSchema).optional(),
  streaks: z.object({ current: z.number(), longest: z.number(), current_start: z.string().nullable().optional() }).optional(),
  eddington: z.array(EddingtonSummarySchema).optional(),
  weekday_stats: z.array(LabeledTotalsSchema).optional(),
  daytime_stats: z.array(LabeledTotalsSchema).optional(),
  distance_breakdown: z.array(LabeledTotalsSchema).optional(),
  hr_zones: z.object({ zones: z.array(z.number()), window_days: z.number() }).nullable().optional(),
  peak_power: z.object({
    durations: z.array(z.number()),
    outputs: z.array(z.object({ duration_s: z.number(), watts: z.number().nullable() })),
    window_days: z.number(),
  }).nullable().optional(),
  training_load: z.array(z.object({ date: z.string(), load: z.number() })).optional(),
  vo2max_trend: z.array(z.object({ date: z.string(), vo2max: z.number() })).optional(),
  training_load_analysis: z.object({
    ctl: z.number(),
    atl: z.number(),
    tsb: z.number(),
    tsb_status: z.string(),
    tsb_color: z.string(),
    ac_ratio: z.number(),
    ac_status: z.string(),
    ac_color: z.string(),
    rest_days: z.number(),
    monotony: z.number(),
    strain: z.number(),
    weekly_trimp: z.number(),
    series: z.array(z.object({
      date: z.string(),
      load: z.number(),
      ctl: z.number(),
      atl: z.number(),
      tsb: z.number(),
      activities: z.array(ActivitySummarySchema.extend({ load: z.number(), intensity: z.number() })).optional(),
    })),
    display_days: z.number(),
  }).nullable().optional(),
  recent_milestones: z.array(MilestoneSchema).optional(),
  gear_stats: z.array(GearItemSchema).optional(),
});

const HrZoneItemSchema = z.object({
  zone: z.number(),
  label: z.string(),
  lower_bpm: z.number(),
  upper_bpm: z.number().nullable(),
  seconds: z.number(),
  percentage: z.number(),
});

const PaceZoneItemSchema = z.object({
  zone: z.number(),
  label: z.string(),
  slow_pace: z.number().nullable(),
  fast_pace: z.number().nullable(),
  seconds: z.number(),
  percentage: z.number(),
});

export const ActivityDetailSchema = ActivitySummarySchema.extend({
  description: z.string().nullable(),
  user_note: z.string().nullable().default(null),
  max_speed_kmh: z.number().nullable(),
  average_cadence: z.number().nullable(),
  max_cadence: z.number().nullable(),
  max_power: z.number().nullable(),
  normalized_power: z.number().nullable(),
  device_name: z.string().nullable(),
  polyline: z.string().nullable(),
  start_latitude: z.number().nullable(),
  start_longitude: z.number().nullable(),
  streams: z.record(z.string(), z.array(z.number().nullable())),
  best_efforts: z.array(z.object({
    distance_m: z.number(),
    label: z.string(),
    time_s: z.number(),
  })),
  hr_zones: z.array(HrZoneItemSchema).nullable().default(null),
  pace_zones: z.array(PaceZoneItemSchema).nullable().default(null),
  hr_curve: z
    .array(z.object({ duration_s: z.number(), bpm: z.number() }))
    .nullable()
    .default(null),
});

export const ImportRunStatusSchema = z.object({
  id: z.number(),
  status: z.string(),
  source: z.string().nullable(),
  added: z.number(),
  updated: z.number(),
  skipped: z.number(),
  deduped: z.number(),
  gear_upserted: z.number(),
  files_parsed: z.number(),
  parse_errors: z.number(),
  total: z.number().nullable(),
  processed: z.number(),
  finished_at: z.string().nullable(),
  message: z.string().nullable(),
});

export const ImportPreviewSchema = z.object({
  source: z.string(),
  provider: z.string(),
  athlete_name: z.string().nullable(),
  source_athlete_id: z.string().nullable(),
  activity_count: z.number(),
  is_existing_athlete: z.boolean(),
  suggested_athlete_id: z.string().nullable(),
  suggested_athlete_name: z.string().nullable(),
});

export const SyncConfigSchema = z.object({
  provider: z.string(),
  athlete_id: z.string(),
  athlete_name: z.string().nullable(),
  icu_athlete_id: z.string(),
  enabled: z.boolean(),
  has_api_key: z.boolean(),
  synced_through: z.string().nullable(),
  last_run_at: z.string().nullable(),
  last_status: z.string().nullable(),
  last_message: z.string().nullable(),
});

export const SyncStatusSchema = z.object({
  configured: z.boolean(),
  enabled: z.boolean(),
  running: z.boolean(),
  synced_through: z.string().nullable(),
  last_run_at: z.string().nullable(),
  last_status: z.string().nullable(),
  last_message: z.string().nullable(),
});

export const SyncRunResultSchema = z.object({
  status: z.string(),
  listed: z.number(),
  added: z.number(),
  updated: z.number(),
  skipped: z.number(),
  deduped: z.number(),
  enriched: z.number(),
  message: z.string().nullable(),
});

// -- Goals ------------------------------------------------------------------

export const GoalResponseSchema = z.object({
  id: z.number(),
  athlete_id: z.string(),
  start_date: z.string(),
  end_date: z.string(),
  sport_type: z.string().nullable(),
  metric: z.string(),
  target_value: z.number(),
  note: z.string().nullable(),
  created_on: z.string(),
  updated_on: z.string(),
});

export const GoalProgressResponseSchema = GoalResponseSchema.extend({
  current_value: z.number(),
  percentage: z.number(),
});

// -- Athlete config ---------------------------------------------------------

export const AthleteConfigSchema = z.object({
  birthday: z.string().nullable(),
  weight_kg: z.number().nullable(),
  ftp: z.number().nullable(),
  max_heart_rate: z.number().nullable(),
  resting_heart_rate: z.number().nullable(),
  unit_system: z.string(),
  threshold_pace: z.number().nullable(),
  heart_rate_zones: z.array(z.number()).nullable(),
  power_zones: z.array(z.number()).nullable(),
  pace_zones: z.array(z.number()).nullable(),
});

File: backend/app/schemas.py
from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel


class ActivitySummary(BaseModel):
    activity_id: str
    name: str
    start_date_time: datetime
    sport_type: str
    sport_label: str
    activity_type: str
    distance_km: float
    distance_mi: float
    elevation_m: float
    moving_time_s: int
    elapsed_time_s: int
    average_speed_kmh: float | None
    average_pace_s_per_km: float | None
    pace_unit: str
    is_distance_based: bool
    average_heart_rate: int | None
    max_heart_rate: int | None
    average_power: int | None
    calories: int | None
    is_commute: bool
    gear_name: str | None
    has_map: bool


class HrZoneItem(BaseModel):
    zone: int
    label: str
    lower_bpm: int
    upper_bpm: int | None
    seconds: int
    percentage: float


class PaceZoneItem(BaseModel):
    zone: int
    label: str
    slow_pace: float | None
    fast_pace: float | None
    seconds: int
    percentage: float


class HrCurvePoint(BaseModel):
    duration_s: int
    bpm: int


class ActivityDetail(ActivitySummary):
    description: str | None
    user_note: str | None = None
    max_speed_kmh: float | None
    average_cadence: int | None
    max_cadence: int | None
    max_power: int | None
    normalized_power: float | None
    device_name: str | None
    polyline: str | None
    start_latitude: float | None
    start_longitude: float | None
    streams: dict[str, list]
    best_efforts: list[BestEffortItem]
    hr_zones: list[HrZoneItem] | None = None
    pace_zones: list[PaceZoneItem] | None = None
    hr_curve: list[HrCurvePoint] | None = None


class BestEffortItem(BaseModel):
    distance_m: int
    label: str
    time_s: float


class PaginatedActivities(BaseModel):
    total: int
    limit: int
    offset: int
    items: list[ActivitySummary]


class GearItem(BaseModel):
    gear_id: str
    name: str
    gear_type: str
    distance_km: float
    is_retired: bool


class ImportRequest(BaseModel):
    source: str
    # ``None`` lets the importer auto-detect the provider from the archive
    # contents (Strava ``activities.csv`` vs Garmin ``DI_CONNECT`` export).
    provider: str | None = None
    force: bool = False
    # When set, merge the import into this existing athlete instead of the
    # provider's own account (``None`` keeps the export's own athlete).
    athlete_id: str | None = None


class ImportPreview(BaseModel):
    """Inspection of an export before importing, used to choose the athlete.

    Lets the dialog show what was detected and offer "import as a new athlete"
    vs "merge into an existing athlete", with a suggested target when the
    provider account was merged before or the name matches an existing athlete.
    """

    # Server-side token to pass back to the import call (stored upload path or
    # the provided server path).
    source: str
    provider: str  # strava | garmin
    athlete_name: str | None = None
    # The provider's own athlete id (Strava athlete id / Garmin userProfileId).
    source_athlete_id: str | None = None
    activity_count: int = 0
    # Whether the provider's own athlete already exists locally (a re-import).
    is_existing_athlete: bool = False
    # Pre-selected merge target (a *different* existing athlete) when one is
    # confidently suggested, else ``None`` (default to a new athlete).
    suggested_athlete_id: str | None = None
    suggested_athlete_name: str | None = None


class ImportRunStatus(BaseModel):
    """Live status of a background import, polled by the client."""

    id: int
    status: str  # running | ok | error
    source: str | None = None
    added: int = 0
    updated: int = 0
    skipped: int = 0
    deduped: int = 0
    gear_upserted: int = 0
    files_parsed: int = 0
    parse_errors: int = 0
    # ``total`` is the number of activities to process (known once the export is
    # read); ``processed`` counts those handled so far. Both drive the progress
    # bar. ``message`` carries an error description when ``status == "error"``.
    total: int | None = None
    processed: int = 0
    finished_at: datetime | None = None
    message: str | None = None


class AthleteListItem(BaseModel):
    athlete_id: str
    name: str | None
    location: str | None
    activity_count: int
    profile_url: str | None


class MetaResponse(BaseModel):
    app_name: str
    app_subtitle: str
    unit_system: str
    distance_unit: str
    elevation_unit: str
    sport_types: list[SportTypeOption]
    activity_count: int
    first_activity: datetime | None
    last_activity: datetime | None
    athlete: AthleteInfo | None = None
    athletes: list[AthleteListItem] = []


class AthleteInfo(BaseModel):
    athlete_id: str | None
    name: str | None
    location: str | None
    profile_url: str | None


class SportTypeOption(BaseModel):
    value: str
    label: str
    activity_type: str


class SyncConfigResponse(BaseModel):
    """Current sync configuration, with the API key redacted.

    ``has_api_key`` tells the UI a key is stored without ever returning it;
    ``athlete_name`` is the display name of the bound canonical athlete.
    """

    provider: str
    athlete_id: str
    athlete_name: str | None = None
    icu_athlete_id: str
    enabled: bool
    has_api_key: bool
    synced_through: datetime | None = None
    last_run_at: datetime | None = None
    last_status: str | None = None
    last_message: str | None = None


class SyncConfigRequest(BaseModel):
    """Create/update the sync configuration. The API key is validated first."""

    athlete_id: str
    api_key: str
    icu_athlete_id: str = "0"
    enabled: bool = True


class SyncTriggerRequest(BaseModel):
    # Ignore the watermark and re-scan from the athlete's earliest anchor.
    full_resync: bool = False


class SyncRunResult(BaseModel):
    """Outcome of a sync run (returned by the trigger endpoint)."""

    status: str  # ok | error
    listed: int = 0
    added: int = 0
    updated: int = 0
    skipped: int = 0
    deduped: int = 0
    enriched: int = 0
    message: str | None = None


class SyncStatusResponse(BaseModel):
    """Pollable last/in-progress sync run state."""

    configured: bool
    enabled: bool = False
    running: bool = False
    synced_through: datetime | None = None
    last_run_at: datetime | None = None
    last_status: str | None = None
    last_message: str | None = None


# -- Activity note ----------------------------------------------------------


class ActivityNoteUpdate(BaseModel):
    note: str | None = None


# -- Goals ------------------------------------------------------------------


class GoalCreate(BaseModel):
    start_date: date
    end_date: date
    # Sports the goal counts toward; an empty list means "all sports".
    sport_types: list[str] = []
    metric: str
    target_value: float
    note: str | None = None


class GoalUpdate(BaseModel):
    start_date: date | None = None
    end_date: date | None = None
    # ``None`` leaves the sports unchanged; a list (including an empty one,
    # meaning "all sports") replaces them.
    sport_types: list[str] | None = None
    metric: str | None = None
    target_value: float | None = None
    note: str | None = None


class GoalResponse(BaseModel):
    id: int
    athlete_id: str
    start_date: date
    end_date: date
    sport_types: list[str]
    metric: str
    target_value: float
    note: str | None
    created_on: datetime
    updated_on: datetime


class GoalProgressResponse(GoalResponse):
    current_value: float
    percentage: float


# -- Athlete config ---------------------------------------------------------


class AthleteConfigResponse(BaseModel):
    birthday: date | None = None
    weight_kg: float | None = None
    ftp: int | None = None
    max_heart_rate: int | None = None
    resting_heart_rate: int | None = None
    unit_system: str = "metric"
    threshold_pace: int | None = None
    heart_rate_zones: list[float] | None = None
    power_zones: list[float] | None = None
    pace_zones: list[float] | None = None


class AthleteConfigUpdate(BaseModel):
    birthday: date | None = None
    weight_kg: float | None = None
    ftp: int | None = None
    max_heart_rate: int | None = None
    resting_heart_rate: int | None = None
    unit_system: str | None = None
    threshold_pace: int | None = None
    heart_rate_zones: list[float] | None = None
    power_zones: list[float] | None = None
    pace_zones: list[float] | None = None


ActivityDetail.model_rebuild()
MetaResponse.model_rebuild()
